client-v2/src/main/java/com/clickhouse/client/api/Client.java

```diff
@@ -159,29 +181,137 @@ public boolean ping(int timeout) {
+            } if (methodName.startsWith("is")) {
+                methodName = methodName.substring(2).toLowerCase();
+                getterMethods.put(methodName, method);
+            }
+        }
+        this.getterMethods.put(clazz, getterMethods);//Store the getter methods for later use
+
+        for (ClickHouseColumn column : schema.getColumns()) {
+            String columnName = column.getColumnName().toLowerCase().replace("_", "");
+            serializers.add((obj, stream) -> {
+                if (!getterMethods.containsKey(columnName)) {
+                    LOG.warn("No getter method found for column: {}", columnName);
+                    return;
+                }
+                Method getterMethod = this.getterMethods.get(clazz).get(columnName);
+                Object value = getterMethod.invoke(obj);
+                boolean hasDefaults = this.hasDefaults.get(clazz);
+
+                //Handle null values
+                if (value == null) {
+                    if (hasDefaults && !column.hasDefault()) {//Send this only if there is no default
+                        BinaryStreamUtils.writeNonNull(stream);
+                    }
+                    BinaryStreamUtils.writeNull(stream);//We send this regardless of default or nullable
+                    return;
+                }
+
+                //Handle default
+                if (hasDefaults) {
+                    BinaryStreamUtils.writeNonNull(stream);//Write 0
+                }
+
+                //Handle nullable
+                if (column.isNullable()) {
+                    BinaryStreamUtils.writeNonNull(stream);//Write 0
+                }
+
+                //Handle the different types
+                SerializerUtils.serializeData(stream, value, column);
+            });
+        }
+        this.serializers.put(clazz, serializers);
+        this.hasDefaults.put(clazz, schema.hasDefaults());
     }
 
     /**
      * Insert data into ClickHouse using a POJO
      */
-    public Future<InsertResponse> insert(String tableName,
-                                     List<Object> data,
-                                     InsertSettings settings,
-                                     List<ClickHouseColumn> columns) throws ClickHouseException, SocketException {
+    public InsertResponse insert(String tableName,
+                                         List<Object> data,
+                                         InsertSettings settings) throws IOException, InvocationTargetException, IllegalAccessException, ExecutionException, InterruptedException {
+        if (data == null || data.isEmpty()) {
+            throw new IllegalArgumentException("Data cannot be empty");
+        }
+        long s1 = System.currentTimeMillis();
+
+        //Add format to the settings
+        if (settings == null) {
+            settings = new InsertSettings();
+        }
+
+        boolean hasDefaults = this.hasDefaults.get(data.get(0).getClass());
```

Comment:
I've just realized that if POJO is not 1:1 mapping of the table then assuming that user what to use defaults may be wrong, so may be make it as settings too?

Problem may be when user mistakenly missed field in a POJO and table has defaults then we would write wrong data actually. 

It is a minor problem.

Reply:
The main problem is to be sure that we write all bytes and did not miss a column because it is missing in pojo.